main.py

```diff
@@ -31,11 +38,232 @@ def setup(self):
+        if not (pipeline.get('status') in {'enabled', 'enabling'}):
+            return
+
+        found_napps = set()
+        content = self.build_content(pipeline)
+        enable_napps = self.get_enabled_napps()
+        # Find NApps in the pipeline to notify them
+        for napp in content:
+            if napp in enable_napps:
+                found_napps.add(napp)
+        self.required_napps = found_napps
+        self.start_enabling_pipeline(content)
+
+    def get_enabled_napps(self) -> set:
+        """Get the NApps that are enabled and subscribed"""
+        enable_napps = set()
+        for key in self.controller.napps:
+            # Keys look like this: ('kytos', 'of_lldp')
+            if key[1] in self.subscribed_napps:
+                enable_napps.add(key[1])
+        return enable_napps
+
+    def start_enabling_pipeline(self, content: dict):
+        """Method to start the process to enable table
+        First, send event notifying NApps about their
+        new table set up.
+        """
+        name = "enable_table"
+        self.emit_event(name, content)
+
+    def build_content(self, pipeline: dict) -> dict:
+        """Build content to be sent through an event"""
+        content = {}
+        for table in pipeline["multi_table"]:
+            table_id = table["table_id"]
+            for napp in table["napps_table_groups"]:
+                if napp not in content:
+                    content[napp] = {}
+                for flow_type in table["napps_table_groups"][napp]:
+                    content[napp][flow_type] = table_id
+        return content
+
+    def emit_event(self, name: str, content: dict = None):
+        """Send event"""
+        context = "kytos/of_multi_table"
+        event_name = f"{context}.{name}"
+        event = KytosEvent(name=event_name, content=content)
+        self.controller.buffers.app.put(event)
+
+    @listen_to("kytos/(mef_eline|coloring|of_lldp).enable_table")
+    def on_enable_table(self, event):
+        """Listen for NApps responses"""
+        self.handle_enable_table(event)
+
+    def handle_enable_table(self, event):
+        """Handle NApps responses from enable_table
+        Second, wait for all the napps to respond"""
+        napp = event.name.split('/')[1].split('.')[0]
+        # Check against the last current table
+        self.required_napps.remove(napp)
+        if self.required_napps:
+            # There are more required napps, 'waiting' responses
+            return
+        self.get_flows_to_be_installed()
+
+    def get_installed_flows(self, pipeline: Dict) -> Optional[Dict]:
+        """Get flows from flow_manager"""
+        command = "v2/stored_flows?state=installed"
+        response = requests.get(f"{FLOW_MANAGER_URL}/{command}")
+
+        if response.status_code // 100 != 2:
+            # Error when enabling pipeline
+            # Return to default pipeline
+            if pipeline.get("status") == "enabling":
```

Comment:
There is no longer roll back. The pipelines will stay in its status and the user can retry with API request on the same pipeline. To avoid conflicts with other pipelines, I restricted other changes to other pipelines until there is not a `enabling` and `disabling` pipeline.

Reply:
Good call. It minimizes edge cases and it's simpler to maintain. Let's close this thread.